### PR TITLE
Rename track slot enumerations

### PR DESCRIPTION
**Pull request by sethrj**
This tries to better distinguish between the "reindexing" we tried with @esseivaju 's work (#717, #727, #765, #1044) and the actual layout changes due to initialization from @amandalund 's (#1322). The I/O layer includes backward compatibility for the slot options.

### CODE REVIEW

File: src/celeritas/Types.hh [hunk 1]
```diff
@@ -128,20 +128,43 @@ enum class StepPoint
...
+ * 1. No sorting is performed and new tracks are inserted in the nearest empty
+ *    track slot. (\c none )
+ * 2. The location of new track slots is biased during track initialization:
+ *    charged and neutral tracks are inserted at opposite sides of the track
+ *    slot vacancies. (\c init_charge )
+ * 3. Tracks are \em reindexed one or more times per step so that the layout
+ *    in memory is unchanged but an additional indirection maps threads onto
+ * different track slots based on particle attributes (\c reindex_status, \c
+ * reindex_particle_type ), actions (\c reindex_along_step_action, \c
+ * reindex_step_limit_action, \c reindex_action ).
```
Comment (amandalund) [inline src/celeritas/Types.hh]:
```suggestion
 * reindex_step_limit_action, \c reindex_both_action ).
```

File: src/celeritas/Types.hh [hunk 2]
```diff
@@ -128,20 +128,43 @@ enum class StepPoint
...
-    shuffled,  //!< Shuffle at the start of the simulation
-    partition_status,  //!< Partition by status at the start of each step
-    sort_along_step_action,  //!< Sort only by the along-step action id
-    sort_step_limit_action,  //!< Sort only by the step limit action id
-    sort_action,  //!< Sort by along-step id, then post-step ID
-    sort_particle_type,  //!< Sort by particle type
-    size_
+    none,  //!< Don't do any sorting: tracks are in an arbitrary order
+    begin_layout_,
+    //! Partition data layout of new tracks by charged vsneutral
```
Comment (amandalund) [inline src/celeritas/Types.hh]:
```suggestion
    //! Partition data layout of new tracks by charged vs neutral
```

File: src/celeritas/global/ActionInterface.hh [hunk 3]
```diff
@@ -57,32 +57,32 @@ class [[deprecated]] ExplicitCoreActionInterface
...
-           || (action == StepActionOrder::along
-               && track == TrackOrder::sort_along_step_action)
-           || (track == TrackOrder::sort_action
-               && (action == StepActionOrder::post
-                   || action == StepActionOrder::along));
+    // CAUTION: check that this matches \c SortTracksAction::SortTracksAction
+    return (aorder == StepActionOrder::post
+            && torder == TrackOrder::reindex_along_step_action)
+           || (aorder == StepActionOrder::along
+               && torder == TrackOrder::reindex_step_limit_action)
```
Comment (amandalund) [inline src/celeritas/global/ActionInterface.hh]: DId the along-step and step limit track orders get swapped here?

**Review by amandalund — COMMENTED** [review]
Looks good @sethrj!

I'm seeing a couple test failures locally:
```console
[ RUN      ] TestEm3DiagnosticTest.device
status: Celeritas core setup complete
info: Executing actions with additional debug checking
status: Celeritas core state initialization complete
/home/alund/celeritas_project/celeritas/test/celeritas/user/Diagnostic.test.cc:242: Failure
JSON objects differ:
  value in .["actions"][0][5]: expected 0, but got 3
  value in .["actions"][0][7]: expected 3, but got 0
  value in .["actions"][0][18]: expected 0, but got 9
  value in .["actions"][0][20]: expected 9, but got 0
  value in .["actions"][1][1]: expected 0, but got 996
  value in .["actions"][1][3]: expected 996, but got 0
  value in .["actions"][1][4]: expected 0, but got 2
  value in .["actions"][1][6]: expected 2, but got 0
  value in .["actions"][1][9]: expected 0, but got 20
  value in .["actions"][1][10]: expected 0, but got 509
  value in .["actions"][1][11]: expected 20, but got 0
  value in .["actions"][1][12]: expected 509, but got 0
  value in .["actions"][1][18]: expected 0, but got 521
  value in .["actions"][1][20]: expected 521, but got 0
  value in .["actions"][2][1]: expected 0, but got 902
  value in .["actions"][2][3]: expected 902, but got 0
  value in .["actions"][2][4]: expected 0, but got 10
  value in .["actions"][2][6]: expected 10, but got 0
  value in .["actions"][2][8]: expected 0, but got 9
  value in .["actions"][2][9]: expected 0, but got 20
  value in .["actions"][2][10]: expected 9, but got 577
  value in .["actions"][2][11]: expected 20, but got 0
  value in .["actions"][2][12]: expected 577, but got 0
  value in .["actions"][2][18]: expected 0, but got 518
  value in .["actions"][2][20]: expected 518, but got 0
/*** ACTUAL ***/
R"json({"_category":"result","_index":["particle","action"],"_label":"action-diagnostic","actions":[[0,0,0,0,0,3,0,0,0,0,0,0,0,0,0,0,0,0,9,0,0,0,0],[0,996,0,0,2,0,0,0,0,20,509,0,0,0,0,0,0,0,521,0,0,0,0],[0,902,0,0,10,0,0,0,9,20,577,0,0,0,0,0,0,0,518,0,0,0,0]]})json"
/******/
Writing diagnostic output because test failed
info: Wrote output to em3diagnostic-device-1.out.json
[  FAILED  ] TestEm3DiagnosticTest.device (16 ms)
```
and 
```console
[ RUN      ] TestEm3SlotTest.device
status: Celeritas core setup complete
info: Executing actions with additional debug checking
status: Celeritas core state initialization complete
/home/alund/celeritas_project/celeritas/test/celeritas/user/SlotDiagnostic.test.cc:259: Failure
Values in: result.slots
 Expected: expected_slots
5 of 16 elements differ
 i   expected_slots     result.slots
11 "000000000000000000000000000000000000000-+----++----+---+-+-+-+-+" "00000000000000000000000000000000000000--+----++----+---+-+-+-+-+"
12 "0000000000000000000000000000000000-0000-+----++----+---+-+-+-+-+" "00000000000000000000000000000000000000--+----++----+---+-+-+-+-+"
13 "0000000000000000000000000-00000-00-0000-+----++--------+-+-+-+-+" "000000000000000000-000000-000000000000--+----++--------+-+-+-+-+"
14 "0000000000000000000000000-0000000000000-+----++--------+-+-+-+-+" "000000000000000000000000000000000000000-+----++--------+-+-+-+-+"
15 "0000000000000000000000000-000000-00-000------++--------+-+-+-+-+" "0000000000000000000000000000000--0-000-------++--------+-+-+-+-+"

Writing diagnostic output because test failed
info: Wrote output to em3slot-device-2.out.json
[  FAILED  ] TestEm3SlotTest.device (36 ms)
```

File: src/celeritas/track/SortTracksAction.cc [hunk 4]
```diff
@@ -61,25 +47,27 @@ SortTracksAction::SortTracksAction(ActionId id, TrackOrder track_order)
...
-    CELER_VALIDATE(is_sort_trackorder(track_order_),
+    CELER_VALIDATE(is_action_sorted(track_order_),
                    << "track ordering policy '" << to_cstring(track_order)
                    << "' should not sort tracks");
-    CELER_EXPECT(track_order != TrackOrder::sort_action);
-    action_order_ = [track_order = track_order_] {
+    CELER_EXPECT(track_order != TrackOrder::reindex_both_action);
+
+    // CAUTION: check that this matches \c is_action_sorted
+    action_order_ = [track_order] {
```
Comment (esseivaju) [inline src/celeritas/track/SortTracksAction.cc:56]: I wonder if there is a clean way to unify these so that they don't need to be kept in sync

**Review by esseivaju — COMMENTED** [review]
I also see `TestEm3DiagnosticTest.device` failing locally with the same output as Amanda

File: src/celeritas/track/SortTracksAction.cc (hunk 4 above)
Comment (sethrj) [reply src/celeritas/track/SortTracksAction.cc:56]: Yeah, I unified one of the two functions but didn't address this one...

**Comment by sethrj** [comment]
Good news, both the failing tests were due to ~~#1433~~ #1426. Fixing now.

**Review by amandalund — APPROVED** [review]
You mean from #1426, I hope? Thanks @sethrj!

**Pull request merged**